tools/fast_build_tar/buildtar.go

```diff
@@ -0,0 +1,370 @@
+package main
+
+import (
+	"archive/tar"
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime/pprof"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+
+	"golang.org/x/build/pargzip"
+)
+
+type tarFile struct {
+	directory string
+
+	tw *tar.Writer
+
+	closers []func()
+}
+
+func newTarFile(output, directory, compression string) (*tarFile, error) {
+	var (
+		w       io.Writer
+		closers []func()
+	)
+	f, err := os.Create(output)
+	if err != nil {
+		return nil, err
+	}
+	closers = append(closers, func() {
+		f.Close()
+	})
+	w = f
+
+	buf := bufio.NewWriter(w)
+	closers = append(closers, func() { buf.Flush() })
+	w = buf
+
+	switch compression {
+	case "":
+	case "gz":
+		gzw := pargzip.NewWriter(w)
+		closers = append(closers, func() { gzw.Close() })
+		w = gzw
+	case "bz2":
+	default:
+		return nil, fmt.Errorf("unknown compression %q", compression)
+	}
+
+	return &tarFile{
+		directory: directory,
+		tw:        tar.NewWriter(w),
+		closers:   closers,
+	}, nil
+}
+
+//TODO(): add ownership
+func (f *tarFile) addFile(file, dest string, mode os.FileMode) error {
+	dest = strings.TrimLeft(dest, "/")
+	dest = filepath.Join(strings.TrimLeft(f.directory, "/"), dest)
+
+	gid := 0
+	uid := 0
+	uname := ""
+	gname := ""
+
+	info, err := os.Stat(file)
+	if err != nil {
+		return err
+	}
+	header := &tar.Header{
+		Name:       dest,
+		Mode:       int64(mode),
```

Comment:
TODO: fix mode based on info?

Reply:
This matches the bazel/build_tar.py behavior. I'm trying to keep these compatible, at least initially.